---
title: "B3_Quality_control"
author: "Alecia-Jane Twigger"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

**Setting up**

*Load Data & libraries*

Load the .rds file, which contains several objects which can be identified by looking at summary().

In the case of ExpressionList.rds, this file contains 2 data frames (phenoData and featureData) and a single matrix (counts). For ease of analysis, it is best to save each of the objects separately by assigning them different names such as pD for phenoData, fD for featureData and m for counts matrix. pD and fD were user generated and provide information about the samples and genes respectively. The counts matrix is direct output from CellRanger software. I compiled these data into an sce object.

```{r, message=FALSE}
#Load libraries
library(scran)
library(scater)
library(dplyr)
library(plyr)
library(knitr)
library(ggplot2)
library(Rtsne)
library(cowplot)
library(Matrix)

#Load data
dataList <- readRDS("../../data/Batch3/ExpressionList.rds")
pD <- dataList$phenoData
fD <- dataList$featureData
m <- as(dataList$counts,"dgCMatrix")
rm(dataList)

#Generate a sce object
sce <- SingleCellExperiment(list(counts=as.matrix(m)),
    colData=DataFrame(pD),
    rowData=DataFrame(fD))

#Add lane info
colData(sce)$Lane <- colData(sce)$Sample

#Mapping samples so they will fit with all merged data
sample.old <- c("HMC","RB1", "RB2","RB3")
sample.new <- c("HMC9","RB8","RB8","RB8")
colData(sce)$Sample <- mapvalues(colData(sce)$Sample,sample.old, sample.new)

state.old <- c("HMC","RB")
state.new <- c("HM","RB")
colData(sce)$State <- mapvalues(colData(sce)$State,state.old, state.new)
```

*Examining the structure of the data*

```{r}
sce
```

*Calculating the total number of UMI's and genes/cell*

Determining the number of UMI's and genes per cell is an important step in quality control

```{r}
sce <- addPerCellQC(sce, subsets=list(Mito=rowData(sce)$Mitochondrial))
colData(sce)
```
**Pre-filter plots**

*Visualizing the number of UMIs and genes detected in each droplet prior to subsetting*

Here we had to convert the colData(sce) into a data frame (pD) so that it can be visualized with ggplots

Using different kinds of plots, we can visualize whether the cells in our data set are of high enough quality by plotting: 
1. the library size which shows the number of UMI's expressed by cells within each sample
2. the number of genes in cells within each sample
3. the number of UMI and genes within the same cell
4. comparison of % mitochondrial counts and UMI per cell
5. comparison of % mitochondrial counts and genes per cell
```{r}
pD <- as.data.frame(colData(sce))

#Adding colour palette
sample.set <- c("HMC9","RB8")
color.names <- c("peachpuff3","lightblue1")

#1. Violin plot of UMI counts/cell per sample
LibrarySize <- ggplot(pD, aes(x=Sample,y=sum, fill=Sample)) +
    geom_violin(draw_quantiles=0.5)+
    scale_y_log10() +
    ylab("Total number of molecules") +
    ggtitle("Number of Molecules") +
    theme_bw() + scale_fill_manual(values=setNames(color.names, sample.set))
LibrarySize
pdf("../../data/Batch3/QualityControl_graphs/UMI_counts.pdf", width=12, height=8)
LibrarySize
dev.off()

#2. Violin plot of genes/cell per sample
genesDetected <- ggplot(pD, aes(x=Sample, y=detected, fill=Sample)) +
  geom_violin(draw_quantiles=0.5) +
  scale_y_log10() +
  ylab("Total number of genes detected") +
  ggtitle("Number of genes detected") +
  theme_bw() + scale_fill_manual(values=setNames(color.names, sample.set))
genesDetected
pdf("../../data/Batch3/QualityControl_graphs/Gene_count.pdf", width=12, height=8)
genesDetected
dev.off()

#3. Scatter plot showing UMI count and gene count per cell
UmiGeneSum <- ggplot(pD, aes(x=sum, y=detected, color=subsets_Mito_percent)) + geom_point(shape=1) + scale_color_gradient(low="gray47", high="red")
UmiGeneSum
pdf("../../data/Batch3/QualityControl_graphs/UmiGeneSum_highUMI.pdf", width=12, height=8)
UmiGeneSum
dev.off()

#4. Coloured scatter plot with mitochondrial counts compared to number of genes colored by UMI count per cell
cellViability <- ggplot(pD, aes(subsets_Mito_percent, detected, color=sum)) + geom_point(alpha=0.5) + 
  scale_color_gradient(low="gray47", high="Blue") +
    xlab("Percentage of mitochondrial reads") +
    ggtitle("Cell Viability")
cellViability
pdf("../../data/Batch3/QualityControl_graphs/Cell_viability.pdf", width=12, height=8)
cellViability
dev.off()

#5. Scatter plot with UMI count compared to % mitochondrial reads
MitoUMI <- ggplot(pD, aes(x=subsets_Mito_percent, y=sum))+
    geom_point(shape=1, alpha=1/2) +
    xlab("Percentage of mitochondrial reads") +
    ggtitle("%Mito vs. UMIs")
MitoUMI
pdf("../../data/Batch3/QualityControl_graphs/Mitocounts_vs_UMIs.pdf", width=12, height=8)
MitoUMI
dev.off()
```
**Filtering the data**
Examining the cells number of UMI, gene and mitochondrial counts, I found that any cells with UMI < 1000 were unlikely to be cells. Therefore these were filtered out together with cells that had cells with large mitochondrial counts.

```{r}
#Filtering cells with low UMI or high mitochondrial counts
mt.outlier <- isOutlier(sce$subsets_Mito_percent, nmads = 1, type="higher")
umi.outlier <- colData(sce)$sum < 1000
colData(sce)$keep <- !mt.outlier & !umi.outlier

#keeping only cells that should not be discarded
sce.pass <- sce[,colData(sce)$keep]
pD.pass <- as.data.frame(colData(sce.pass))
```
*Examining the kept vs. discarded cells in more detail*
```{r}
#Plotting kept vs. discarded cells
gridExtra::grid.arrange(
    plotColData(sce, x="Sample", y="sum", colour_by="keep",
        other_fields="State") + facet_wrap(~State) + 
        scale_y_log10() + ggtitle("Total count"),
    plotColData(sce, x="Sample", y="detected", colour_by="keep", 
        other_fields="State") + facet_wrap(~State) + 
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(sce, x="Sample", y="subsets_Mito_percent", 
        colour_by="keep", other_fields="State") + 
        facet_wrap(~State) + ggtitle("Mito percent"),
    ncol=1
)

#Looking to see whether discarded pool contains highly expressed marker genes for a specific cell type. Mitochondrial genes are in blue.
library(edgeR)
lost <- calculateAverage(counts(sce)[,!colData(sce)$keep])
kept <- calculateAverage(counts(sce)[,colData(sce)$keep])
logged <- cpm(cbind(lost, kept), log=TRUE, prior.count=2)
logFC <- logged[,1] - logged[,2]
abundance <- rowMeans(logged)
plot(abundance, logFC, xlab="Average count", ylab="Log-FC (lost/kept)", pch=16)
points(abundance[rowData(sce)$Mito], logFC[rowData(sce)$Mito], col="dodgerblue", pch=16)
```
*Re-visualising all cells but now looking at only the kept cells*
```{r}
#visualising all the kept cells
LibrarySize %+% pD.pass
genesDetected %+% pD.pass
UmiGeneSum %+% pD.pass
cellViability %+% pD.pass
MitoUMI %+% pD.pass

pdf("../../data/Batch3/QualityControl_graphs/Gene_count_clean.pdf", width=12, height=8)
genesDetected %+% pD.pass
dev.off()

pdf("../../data/Batch3/QualityControl_graphs/UMI_count_clean.pdf", width=12, height=8)
LibrarySize %+% pD.pass
dev.off()

pdf("../../data/Batch3/QualityControl_graphs/UmiGeneSum_clean.pdf", width=12, height=8)
UmiGeneSum %+% pD.pass
dev.off()

pdf("../../data/Batch3/QualityControl_graphs/Cell_viability_clean.pdf", width=12, height=8)
cellViability %+% pD.pass
dev.off()

pdf("../../data/Batch3/QualityControl_graphs/Mitocounts_vs_UMIs_clean.pdf", width=12, height=8)
MitoUMI %+% pD.pass
dev.off()
```
*Summary tables of cell pre and post filter*
```{r}
#Pre-filter cell summary
sumry1 <- group_by(pD, Sample) %>%
    dplyr::summarize("Total Number of cells_pre-filter"=n(),
              "Total molecules_pre-filter"=median(sum),
              "Genes Detected_pre-filter"=median(detected))
kable(sumry1,caption="Number of cells pre-filter")
write.csv(sumry1, "../../data/Batch3/QualityControl_graphs/Pre-filter_summary.csv")

#Post-filter cell summary
sumry <- group_by(pD.pass, Sample) %>%
    dplyr::summarize("Number of cells"=n(),
          "Total molecules"=median(sum),
          "Genes Detected"=median(detected))
kable(sumry,caption="Summary of QC parameters as median")
write.csv(sumry, "../../data/Batch3/QualityControl_graphs/Post-filter_summary.csv")
```
**Normalization**

*Only do this once filtering has been done.*
Here we use a method of normalization that was developed by Aaron Lun and Karsten Bach (Lun et al. Genome Biology 2016)

1. Firstly the data needs to be clustered into relatively arbitrary clusters that can be used in the downstream step to normalize the data.
2. Using the command computeSumFactors(), we have deduced the size factor (sf) which will be used to multiply each cell in our matrix by the gene, so that each gene is normalized. We can visualise this by comparing the sf with the Umicounts of the cells. From the graph, we can see that actually these two factors are correlated, however  this is not a perfect correlation.
3. Finally the normalization
```{r}
clusters <- quickCluster(sce.pass, use.ranks="FALSE", method="igraph")
table(clusters)
sce.pass <- computeSumFactors(sce.pass, clusters=clusters)
sf <- plot(log10(sizeFactors(sce.pass)),log10(colData(sce.pass)$sum),pch=19)
sce.pass <- scater::logNormCounts(sce.pass)
```
**Saving the data**
Saving the new data sets that have been filtered and normalized with the new name "sce_clean.rds".
```{r}
saveRDS(sce.pass,file="../../data/Batch3/sce_clean.rds")
sessionInfo()
```